### DISCUSSION TITLE
4.0

### DISCUSSION THREAD

**Discussion by r0x0r** [r0x0r/pywebview#1016] [General]
### 💔 BREAKING CHANGES
- [All] Window events are moved into `window.events` namespace. `window.loaded`, `window.shown` etc no longer work.
- EdgeHTML support is removed.

### ⚡ Features
- [All] Local homegrown HTTP server is replaced with [bottle.py](https://bottlepy.org). Thanks @robb-brown for WSGI support.
- [All] Native application menu support. See `examples/menu.py` for usage example. Thanks @sardination
- [All] `webview.start(private_mode=True, storage_path=None)` Private mode and persistant storage support in a non-private mode. Private mode is enabled by default.
- [All] `webview.create_window(zoomable=False)` Enable / disable zooming on webpage. Disabled by default.
- [All] `webview.create_window(draggable=False)` Enable / disable dragging of IMG and A elements. Disabled by default.
- [All] `webview.create_confirmation_dialog(title, content)` creates a confirmation (Ok, Cancel) dialog. Thanks @sardination.
- [All] `window.get_cookies()` retrieve all the cookies (including HttpOnly) for the current webpage.
- [macOS] `webview.create_window(vibancy=False)` Window vibrancy suppport. macOS only. Thanks @CahierX.

### 🚀 Improvements
- [All] Local relative URLs (eg. src/index.html) are opened using the built-in http server by default. Support for local URLs is still possible using file:// schema
- [Cocoa] Disable Ctrl+click context menu. Thanks @ecpost.
- [EdgeChromium] Improve `evaluate_js` performance.
- [GTK] Enable media / audio / WebGL / clipboard related WebKit features

### 🐞 Bug fixes
- [Cocoa] Fix passing through keyboard events handled by pywebview. Thanks @ecpost.
- [GTK] Fix JS bridge maximum return object size limitation. GTK's JS bridge is implemented via HTTP server.
- [GTK] Fix hanging problem during window closing when JS evaluation is in progress

<hr /><em>This discussion was created from the release <a href='https://github.com/r0x0r/pywebview/releases/tag/4.0'>4.0</a>.</em>

**Comment by ToonJorik** [comment]
Здравствуйте! Не нашел другого способа с вами связаться.
Решил попробовать pywebview для отображения веб страниц внутри программы Python.
Столкнулся с проблемой: окна webview блокируют работу основного окна программы.
Пробовал запустить окно в отдельном потоке - не помогло.
Так и должно быть или это баг? Или я неверно использую webview?
Более подробно проблему описал на [stackoverflow](https://ru.stackoverflow.com/questions/1491490/pywebview-%d0%9a%d0%b0%d0%ba-%d0%bf%d0%b5%d1%80%d0%b5%d0%ba%d0%bb%d1%8e%d1%87%d0%b8%d1%82%d1%8c-%d1%84%d0%be%d0%ba%d1%83%d1%81-%d0%bd%d0%b0-%d0%be%d1%81%d0%bd%d0%be%d0%b2%d0%bd%d0%be%d0%b5-%d0%be%d0%ba%d0%bd%d0%be-%d0%bf%d1%80%d0%be%d0%b3%d1%80%d0%b0%d0%bc%d0%bc%d1%8b)
Буду признателен за ответ.
С уважением, Павел.

**Reply by r0x0r** [reply]
Вы правы, что webview.start() блокирует дальнейшее исполнения кода. Это фича гуй-библиотек, также как и тот факт что гуй должен исполнятся в главном потоке. Архитектура pywebview подразумевает исполнения гуя в главном потоке и своего кода в дополинтельном потоке. Если вам надо исполнять и то и другое в главном потоке, попробуйте  в качестве решения multiprocessing.

**Reply by ToonJorik** [reply]
Спасибо, буду разбираться. Надеюсь мне подойдет, так как инструмент показался очень годным и простым. Совсем не хочется заморачиваться с QtWebengine.

**Reply by ToonJorik** [reply]
Действительно через multiprocessing заработало. Быть может стоит отразить этот нюанс в мануале и примерах?
Вот такое накидал, кривенько, но работает.

import multiprocessing
import sys
import webview

from PyQt5 import QtWidgets
from PyQt5.Qt import *


class Window(QMainWindow):
    def __init__(self):
        super(Window, self).__init__()
        self.resize(400, 200)

        self.centralwidget = QtWidgets.QWidget()
        self.setCentralWidget(self.centralwidget)

        self.label = QtWidgets.QLabel(f'<h1>Click me!</h1>', alignment=Qt.AlignCenter)
        self.label.setStyleSheet('background-color: #5599ff;')

        self.gridLayout = QtWidgets.QGridLayout(self.centralwidget)
        self.gridLayout.addWidget(self.label, 0, 0, 1, 1)

    def mousePressEvent(self, event):
        pass

def gui_main():
    app = QApplication(sys.argv)
    window = Window()
    window.show()
    sys.exit(app.exec_())

def gui_bro():
    win2 = webview.create_window('tst_win_2', html='text2', width=400, height=300)
    webview.start(win2)

if __name__ == '__main__':

    proc1 = multiprocessing.Process(target=gui_main)
    proc2 = multiprocessing.Process(target=gui_bro)
    proc1.start()
    proc2.start()
    proc1.join()
    proc2.join()
